### PR TITLE
Create JSON or YAML Files with Loop Features/Configurations

### PR DESCRIPTION
**Pull request by mostafaelhoushi**
Added an option to the `opt_loops` tool to dump a YAML loop features/configuration file. The objectives of this YAML file are:
- log features of each loop, to provide CompilerGym with per-loop observation
- [in the future] could act as a configuration file to control optimizations of each loop

Here is an example of how the YAML file looked like for [`add.c`](https://github.com/facebookresearch/CompilerGym/blob/development/examples/loop_optimizations_service/benchmarks/add.c) benchmark:

```
---
ID:              '!2'
Function:        add
Module:          '/tmp/add.ll'
MetadataID:      4
Name:            '<unnamed loop>'
Depth:           1
HeaderName:      loop_0
llvm.loop.unroll.enable: true
llvm.loop.unroll.disable: false
llvm.loop.unroll.count: 32
llvm.loop.isunrolled: false
llvm.loop.vectorize.enable: false
llvm.loop.vectorize.disable: false
llvm.loop.isvectorized: false
llvm:            '  Loop at depth 1 containing: 
<header><exiting>
loop_0:                                           ; preds = %19, %1
  %4 = load i32, i32* %3, align 4
  %5 = icmp slt i32 %4, 1000000
  br i1 %5, label %6, label %22


6:                                                ; preds = %loop_0
  %7 = load i32, i32* %3, align 4
  %8 = sext i32 %7 to i64
  %9 = getelementptr inbounds [1000000 x i32], [1000000 x i32]* @A, i64 0, i64 %8
  %10 = load i32, i32* %9, align 4
  %11 = load i32, i32* %3, align 4
  %12 = sext i32 %11 to i64
  %13 = getelementptr inbounds [1000000 x i32], [1000000 x i32]* @B, i64 0, i64 %12
  %14 = load i32, i32* %13, align 4
  %15 = add nsw i32 %10, %14
  %16 = load i32, i32* %3, align 4
  %17 = sext i32 %16 to i64
  %18 = getelementptr inbounds [1000000 x i32], [1000000 x i32]* @A, i64 0, i64 %17
  store i32 %15, i32* %18, align 4
  br label %19

<latch>
19:                                               ; preds = %6
  %20 = load i32, i32* %3, align 4
  %21 = add nsw i32 %20, 1
  store i32 %21, i32* %3, align 4
  br label %loop_0, !llvm.loop !2
'
...
---
ID:              '!5'
Function:        main
Module:          '/tmp/add.ll'
MetadataID:      4
Name:            '<unnamed loop>'
Depth:           1
HeaderName:      loop_1
llvm.loop.unroll.enable: true
llvm.loop.unroll.disable: false
llvm.loop.unroll.count: 32
llvm.loop.isunrolled: false
llvm.loop.vectorize.enable: false
llvm.loop.vectorize.disable: false
llvm.loop.isvectorized: false
llvm:            '  Loop at depth 1 containing: 
<header><exiting>
loop_1:                                           ; preds = %18, %2
  %15 = load i32, i32* %9, align 4
  %16 = icmp slt i32 %15, 100
  br i1 %16, label %17, label %21


17:                                               ; preds = %loop_1
  call void @add(i32* %6)
  br label %18

<latch>
18:                                               ; preds = %17
  %19 = load i32, i32* %9, align 4
  %20 = add nsw i32 %19, 1
  store i32 %20, i32* %9, align 4
  br label %loop_1, !llvm.loop !2
'
...
```

Some issues that we need to fix/improve in the YAML file:
- need to find a unique ID or name for a loop that persists across multiple passes on the LLVM IR file: currently, when I run optimizations like loop unroll or loop vectorize, fields like `name`, `id`, etc. change
- create a single list of YAML entries, rather than a separate YAML list for each loop
- add an option to hide the loops of the `main()` function (or to be generic, hide/show loops of specific functions)

For next steps, I plan to:
- focus on extracting features like AutoPhase, ir2vec, and ProGraml from both the module IR as well as loop IR

### CODE REVIEW

File: examples/loop_optimizations_service/opt_loops/opt_loops.cc [hunk 1]
```diff
@@ -47,6 +149,13 @@ cl::opt<std::string> InputFilename(cl::Positional, cl::desc("Specify input filen
 cl::opt<std::string> OutputFilename("o", cl::desc("Specify output filename"),
                                     cl::value_desc("filename"), cl::init("-"));
 
+/// Output YAML log file name.
+cl::opt<std::string> OutputYAMLFile("emit-yaml", cl::desc("Specify output YAML log filename"),
+                                    cl::value_desc("filename"), cl::init("/tmp/loops.log"));
+
+// TODO: add other features like "read-yaml", "print-yaml-after-all", "print-yaml-before-all",
```
Comment (ChrisCummins) [inline examples/loop_optimizations_service/opt_loops/opt_loops.cc]: TODOs should be blameable, e.g.

```suggestion
// TODO(mostafaelhoushi): add other features like "read-yaml", "print-yaml-after-all", "print-yaml-before-all",
```

or:

```suggestion
// TODO(github.com/facebookresearch/CompilerGym/issues/553): add other features like "read-yaml", "print-yaml-after-all", "print-yaml-before-all",
```

This applies here and elsewhere in this PR.

**Review by ChrisCummins — COMMENTED** [review]
Cool!

> * create a single list of YAML entries, rather than a separate YAML list for each loop

This should probably be fixed prior to merging, as I think this means the current generated YAML is malformed. Do you care about YAML? If not, you could use JSON. There is a nice library for working with JSON in C++. [Here is an example](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/service/Observation.cc#L106-L127) of it being used in CompilerGym.

Or is the idea that this output will be parsed by a python environment to extract the features? If so, you may want to use protobufs, as they provide significantly faster deserialization than JSON/YAML, while still providing a text format for human readable debugging.

> * need to find a unique ID or name for a loop that persists across multiple passes on the LLVM IR file: currently, when I run optimizations like loop unroll or loop vectorize, fields like `name`, `id`, etc. change

Could [LLVM-Canon](https://github.com/michalpaszkowski/LLVM-Canon) help here?

**Comment by mostafaelhoushi** [comment]
I refactored the code and added another option to dump to JSON. So now, `-emit-json` logs to a JSON file while `-emit-json` logs to a YAML file.
I can add an option to log to protobuf.

I also added IRCanonicalizer as a third-party directory, and made it optional to run its pases in our `opt_loops` tool. However, it seems from the first try that it didn't solve the problem. But perhaps I can make a closer look.

> Cool!
> 
> > * create a single list of YAML entries, rather than a separate YAML list for each loop
> 
> This should probably be fixed prior to merging, as I think this means the current generated YAML is malformed. Do you care about YAML? If not, you could use JSON. There is a nice library for working with JSON in C++. [Here is an example](https://github.com/facebookresearch/CompilerGym/blob/development/compiler_gym/envs/llvm/service/Observation.cc#L106-L127) of it being used in CompilerGym.
> 
> Or is the idea that this output will be parsed by a python environment to extract the features? If so, you may want to use protobufs, as they provide significantly faster deserialization than JSON/YAML, while still providing a text format for human readable debugging.
> 
> > * need to find a unique ID or name for a loop that persists across multiple passes on the LLVM IR file: currently, when I run optimizations like loop unroll or loop vectorize, fields like `name`, `id`, etc. change
> 
> Could [LLVM-Canon](https://github.com/michalpaszkowski/LLVM-Canon) help here?

File: compiler_gym/third_party/LLVM-Canon/BUILD [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the LICENSE file
+# in the root directory of this source tree.
+#
+# This package exposes the LLVM optimization pipeline as a CompilerGym service.
```
Comment (ChrisCummins) [inline compiler_gym/third_party/LLVM-Canon/BUILD]:
```suggestion
```

File: compiler_gym/third_party/LLVM-Canon/BUILD [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    name = "IRCanonicalizer",
+    srcs = [
+        "IRCanonicalizer.cc",
+    ],
+    hdrs = [
+        "IRCanonicalizer.h",
+    ],
+    copts = [
+        "-Wall",
+        "-fdiagnostics-color=always",
```
Comment (ChrisCummins) [inline compiler_gym/third_party/LLVM-Canon/BUILD]:
```suggestion
```

not sure these are necessary

File: compiler_gym/third_party/LLVM-Canon/README.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# LLVM-Canon
+LLVM-Canon aims to transform LLVM modules into a canonical form by reordering and renaming instructions while preserving the same semantics.
```
Comment (ChrisCummins) [inline compiler_gym/third_party/LLVM-Canon/README.md]: Could you please extend this to follow [this template](https://github.com/facebookresearch/CompilerGym/tree/development/compiler_gym/third_party/autophase)? Specifically, add an author list to credit @michalpaszkowski, the URL of the upstream project https://github.com/michalpaszkowski/LLVM-Canon, and hash of the git commit this version is based on. Thanks!

File: examples/loop_optimizations_service/opt_loops/opt_loops.cc [hunk 5]
```diff
@@ -138,13 +308,29 @@ class LoopCounter : public llvm::FunctionPass {
...
+
+    for (auto L : Loops) {
+      LoopConfig LC(L);
+      LCs.push_back(LC);
+    }
 
     return false;
   }
+
+ protected:
```
Comment (ChrisCummins) [inline examples/loop_optimizations_service/opt_loops/opt_loops.cc]:
```suggestion
```

File: examples/loop_optimizations_service/opt_loops/opt_loops.cc [hunk 6]
```diff
@@ -182,7 +368,7 @@ char LoopConfiguratorPass::ID = 1;
 /// On error, messages are written to stderr and null is returned.
 ///
 /// \param Context LLVM Context for the module.
-/// \param Name Input file name.
+/// \param Name Input file Name.
```
Comment (ChrisCummins) [inline examples/loop_optimizations_service/opt_loops/opt_loops.cc]:
```suggestion
/// \param Name Input file name.
```

**Review by ChrisCummins — APPROVED** [review]
Hey @mostafaelhoushi, thanks for taking a look at my suggestions and sorry for the extra hassle!

> I also added IRCanonicalizer as a third-party directory, and made it optional to run its pases in our opt_loops tool. However, it seems from the first try that it didn't solve the problem. But perhaps I can make a closer look.
Ah, that's a pity. It will be useful to have the canonicalizer in our codebase, I think it would make a useful addition to the python APIs.

The README needs a few extra details (see inline comment). Aside from that, and some minor nitpicks, LGTM! Please merge when you're happy.

Cheers,
Chris

File: compiler_gym/third_party/LLVM-Canon/README.md (hunk 4 above)
Comment (mostafaelhoushi) [reply compiler_gym/third_party/LLVM-Canon/README.md]: Done

**Comment by mostafaelhoushi** [comment]
Thanks Chris. There was no hassle at all :) 
I have also made sure that it builds with CMake on Linux

**Pull request merged**